src/java/org/apache/cassandra/guardrails/ValueBasedGuardrail.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.guardrails;
+
+import javax.annotation.Nullable;
+
+import org.apache.cassandra.service.QueryState;
+
+/**
+ * A guardrail based on two predicates.
+ *
+ * <p>A {@link ValueBasedGuardrail} guardrail defines a basic guardrail that may warn or fail,
+ * depending on provided value
+ *
+ * @param <T> the type of the values to be tested.
+ */
+public interface ValueBasedGuardrail<T> extends Guardrail
```

Comment:
I found this name confusing, as it's very close to `ValuesBasedGuardrail` which is something different. Could it not be something like `PredicateBasedGuardrail`?